File: Forge/src/main/java/hantonik/atomic/core/utils/AtomicShapeBuilder.java
package hantonik.atomic.core.utils;

import com.google.common.collect.Maps;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;

import java.util.Map;
import java.util.function.Function;

public class AtomicShapeBuilder {
    private static final Map<BlockState, VoxelShape> SHAPES = Maps.newHashMap();

    private VoxelShape leftShape;
    private VoxelShape lastOrShape;

    public static VoxelShape getOrCreate(BlockState state, Function<BlockState, VoxelShape> shapeFactory) {
        return SHAPES.computeIfAbsent(state, shapeFactory);
    }

    public static AtomicShapeBuilder fromShapes(VoxelShape... shapes) {
        var builder = new AtomicShapeBuilder();

        for (var shape : shapes)
            builder.shape(shape);

        return builder;
    }

    public AtomicShapeBuilder shape(VoxelShape shape) {
        if (this.leftShape == null)
            this.leftShape = shape;

        else {
            var newShape = Shapes.or(this.leftShape, shape);

            if (this.lastOrShape != null)
                this.lastOrShape = Shapes.or(this.lastOrShape, newShape);

            else
                this.lastOrShape = newShape;

            this.leftShape = null;
        }

        return this;
    }

    public VoxelShape build() {
        return this.lastOrShape;
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/util/AtomicShapeBuilder.kt
package hantonik.atomic.core.util

import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape

class AtomicShapeBuilder {
    private var leftShape: VoxelShape? = null
    private var lastOrShape: VoxelShape? = null

    companion object {
        private val SHAPES: MutableMap<BlockState, VoxelShape> = hashMapOf()

        fun getOrCreate(state: BlockState, shapeFactory: (BlockState) -> VoxelShape): VoxelShape = SHAPES.computeIfAbsent(state, shapeFactory)

        fun fromShapes(vararg shapes: VoxelShape): AtomicShapeBuilder {
            val builder = AtomicShapeBuilder()

            for (shape in shapes)
                builder.shape(shape)

            return builder
        }
    }

    fun shape(shape: VoxelShape): AtomicShapeBuilder {
        if (this.leftShape == null)
            this.leftShape = shape
        else {
            val newShape = Shapes.or(this.leftShape!!, shape)

            if (this.lastOrShape != null)
                this.lastOrShape = Shapes.or(this.lastOrShape!!, newShape)
            else
                this.lastOrShape = newShape

            this.leftShape = null
        }

        return this
    }

    fun build(): VoxelShape = this.lastOrShape!!
}